### DISCUSSION TITLE
Templater, Dataview Api

### DISCUSSION THREAD

**Discussion by YukiGasai** [YukiGasai/obsidian-google-calendar#110] [General]
## What do we want
As suggested by @rapatel0 in #9 the plan is to expose an API to use the Google calendar results inside dataview and templater.

I created a first test.
The syntax can look something like this:

## Simple Dataview test
```
    ```dataviewjs
     const {getEvents} = this.app.plugins.plugins["google-calendar"].api
     const events = await getEvents();
     dv.list(events);
    ```
```
## Simple Templater test
```
<%*
const {getEvents} = this.app.plugins.plugins["google-calendar"].api;

const events = await getEvents();
const eventsString = events.reduce((text, event) => text +=`- ${event.summary} \n`, "");

tR = `
Todays events: 
${eventsString}
`
%>
```

## Advanced example
```
<%*
//Get function from Api
const {getEvents} = this.app.plugins.plugins["google-calendar"].api;
//Get all events from day
const events = await getEvents(moment("09.29.2022"));
//Filter to get a single event
const theEvent = events.find(event=>event.summary === "She Hulk");
//Map attendees to a singel string
const attendees = theEvent.attendees.reduce((text,attendee) => text+= `[[${attendee.email}]] `, "");

//Display 
tR = `
Attendees: ${attendees}
`%>

Some more templater stuff
```

Currently, the API supports the following functions:
```
getCalendars: () => Promise<GoogleCalendar[]>
getEvent: (id:string, calendarId:string) => Promise<GoogleEvent>
getEvents: (start?:moment.Moment, end?:moment.Moment) => Promise<GoogleEvent[]>
getEventsFromCalendar: (calendar:GoogleCalendar, start?:moment.Moment, end?:moment.Moment) => Promise<GoogleEvent[]>
```

TODO:
- [x] Allow to create, update, maybe delete event from API
- [ ] Document in the README

Any feedback is appreciated. 😊

**Comment by YukiGasai** [comment]
Updating the example:

~~~
Example:
<%*
const {getEvents} = this.app.plugins.plugins["google-calendar"].api;
const theEvents = await getEvents({
startDate: window.moment("12/06/2022")
});
console.log(theEvents);
tR = theEvents.reduce((text, event )=> text += "\n" + event.summary,  "");
%>
~~~

The argument for the getEvents function is a object now to allow named params
```
{
    startDate?: moment.Moment;        Querry Start defaults to start of today
    endDate?: moment.Moment;          Querry End   defaults to end of today
    singleCalendar?: GoogleCalendar;  A single calendar not realy practical to use in templater
    exclude?: string[];               A list of names or ids of calendars you want to exclue from the results defaults to empty
    include?: string[];               A list of names or ids of calendars you want to include in the results  defaults to empty
}
```
when not using include or exclude, the result will contain all your calendars except the blacklisted ones from the settings page

**Reply by barnabaslabs** [reply]
Hi, 
I'm trying to use dataview.js to pull the current event information into my event note. For example, I have an event named "Meeting with Ben." My events often get more details added to them as they get closer, so I'm using dataviewjs to pull the info dynamically. Perhaps I need to use a combination of dataview and templater to achieve what I want.

Here's part of the template:
```dataviewjs
     const {getEvent} = this.app.plugins.plugins["google-calendar"].api
     const frontmatter = dv.current("event-id");
     const event = await getEvent('"' + dv.current("event-id") + '"',"myemail@gmail.com");
     console.log(event);
const startDate = new Date(event.start.dateTime); 
const formattedStartDate = startDate.toLocaleDateString('en-US', { year: 'numeric', month: '2-digit', day: '2-digit' });
	 dv.header(1,event.summary);
	 dv.paragraph(event.description);
	 dv.paragraph(formattedStartDate);
	 dv.paragraph(event.location);
```
The problem I'm having is pulling the event-id from the frontmatter. I was trying to do it with this `const frontmatter = dv.current("event-id");`. But it doesn't seem to work. How could I pass the event-id into the template when the note is created?

**Comment by VladaSedlacek** [comment]
Hi, thanks for adding this functionality! 
I'm not sure if this is a relevant place to post this, but I tweaked your example a little so that it works with the [Day Planner](https://github.com/lynchjames/obsidian-day-planner) plugin and thought someone might find it useful. Namely, if you add
```
<%*
const {getEvents} = this.app.plugins.plugins["google-calendar"].api;
const theEvents = await getEvents({
startDate: moment(tp.file.title)
});
//pull events, highlight them and add check-boxes; all-day events default to 12:00
tR += theEvents.reduce((text, event )=> text += "\n- [ ] " + ((moment(event.start.dateTime).format("HH:mm") == tp.date.now("HH:mm"))? "12:00" : moment(event.start.dateTime).format("HH:mm")) + " ==" + event.summary + "==", "");
%>
```
to your day schedule template, all events on that day are automatically pulled from the calendar in the right format. This works even if you create the note too early, as long as your note name format is supported by [Moment.js](https://momentjs.com/docs/#/parsing/); otherwise you can just replace `moment(tp.file.title)` with `tp.date.now("YYYY-MM-DD")`.

**Comment by imed-ghomari** [comment]
Hi, thank you for this plugin. I want to use this API with the templater plugin, but I didn't find any documentation. Here is my use cases, I want to write 2 scripts : 
1. When I create a note, the plugin automatically creates an event with the same name in the default calendar set in the settings, I can add the date in the scripts.
2. When I view this same note and delete it, it also deletes the corresponding event that was previously created with the firs script.

> - [x] Allow to create, update, maybe delete event from API

As it's written here, there already are functions to create and delete. So can you give me some pointers on how to accomplish this ?

PS, If you think it'll be useful, I'd be happy to write a step-by-step guide on how to accomplish my use cases in your readme.

**Comment by YukiGasai** [comment]
Hi @imeed166,
take a lot [here](https://yukigasai.github.io/obsidian-google-calendar/#/Api/Overview), feel free to improve this documentation based on your experience with it. I would really appreciate that.

**Comment by imed-ghomari** [comment]
This is what I found : 

You could add an example dataview script for the read function (if possible, return only the event title and group by start date) and clearly differentiate between a templater script and a dataview script in the examples.

For the update and delete function, you could a templater example to target the current note, but I realize this is for my specific use case, so might not be appropriate as a general example.

**Comment by YukiGasai** [comment]
Thank you for your feedback.
I hope you were able to solve your problem.
I will update the documentation, but can accept your pull requests because they are not adding enough information.

**Comment by YukiGasai** [comment]
The API will now throw GoogleAPIErrors if there are problems with the request. This was recommended in #90. 
To handle this errros your can surrond your code with try catch.

Example:

~~~
```dataviewjs
   const {getEvents} = this.app.plugins.plugins["google-calendar"].api
   try{
      const events = await getEvents({
         startDate: window.moment("03/20/2023"),
         endDate: window.moment("03/22/2023")
      });
      dv.list(events);  
   }catch(error){
      console.log({error});
   }
```
~~~

**Comment by califa** [comment]
What's the easiest way to get the current event in Templater as exposed in `gEvent` for automatic meeting notes? 
Seems like I need both the event id and the cal id but not sure how to get those for the specific event.

**Reply by YukiGasai** [reply]
I am not exactley sure what you mean but here is a example that inserts most of the info.
~~~
<%*
const {getEvent} = this.app.plugins.plugins["google-calendar"].api;  
const event = await getEvent("{{gEvent.id}}", "{{gEvent.parent.id}}");
%>
# <%event.summary%>

## Meeting Metadata
### Date
Start: <%
event.start.date 
? window.moment(event.start.date).format("DD-MM-YYYY") 
: window.moment(event.start.dateTime).format("DD-MM-YYYY hh:mm")
%>
End: <%
event.start.date 
? window.moment(event.end.date).format("DD-MM-YYYY") 
: window.moment(event.end.dateTime).format("DD-MM-YYYY hh:mm")
%>

### Attendees 
| Name | Status |
| --- | --- |
<%
tr = event?.attendees?.reduce((all,attendee) => `${all}|${attendee.displayName??attendee.email}|${attendee.responseStatus}|\n`,'') ?? ""
%>

### Location
<%event.location ?? ""%>
<%event.hangoutLink ?? ""%>

### Description
<%event.description ?? "No details"%>

### Attachments
<%
tr = event?.attachments?.reduce((all,attachment) => `${all}- [${attachment.title}](${attachment.fileUrl})\n`,'') ?? ""
%>
~~~

**Reply by califa** [reply]
This is great, thank you so much!
